### PR TITLE
Skin shader example update.

### PR DESCRIPTION
**Pull request by tschw**
Includes #7109 . 

I wasn't sure whether the derivatives approach is the preferred way of fixing it. In case not, you can use this one for testing and merge the amendment of the incomplete lighting update. For all changes, just merge this one.

### CODE REVIEW

File: examples/js/ShaderSkin.js [hunk 1]
```diff
@@ -167,7 +167,7 @@ THREE.ShaderSkin = {
 				"diffuseColor = diffuseColor * colDiffuse;",
 
 				"vec3 normal = normalize( vNormal );",
-				"vec3 viewPosition = normalize( vViewPosition );",
+				"vec3 viewerDirection = normalize( vViewPosition );",
```
Comment (WestLangley) [inline examples/js/ShaderSkin.js]: We have used `viewDir` and `viewVector` elsewhere in the library for this term. It looks like the latter one would most consistent with the source code in this PR.

**Comment by WestLangley** [comment]
Can you explain what the pixelSize addition is all about?

How do the original tangent approach, the new computed tangent approach, and the derivative tangent approach compare visually in your opinion?

**Comment by tschw** [comment]
> Can you explain what the pixelSize addition is all about?

The basic idea is differentiation of one coordinate component of the UVs:

We easily get the U gradient in UV-space units in respect to screen space, but it's incomplete for the lack of a "`dFdz`". Normalization yields the screen-projected tangent, getting rid of the useless UV-space unit.

Now we still need a Z component. Therefore I use the directional derivative of Z. This one comes in view space units in respect to screen space.

Now `pixelSize`, the view space size of a pixel, comes into play to match the 2D tangent to the same unit as the Z added.

> How do the original tangent approach, the new computed tangent approach, and the derivative tangent approach compare visually in your opinion?

What's "new computed tangent approach"? I might have just missed it... Should the warning message ",computeTangents() has been removed" (cited from memory) perhaps contain a hint where to find it?

As far as this example is concerned, the normal mapping looks better than before. I guess it's due to the construction of a properly orthogonalized transform (final tangent := binormal x normal). That part does not require the tangent to be obtained in a particular way, though.

I came up with the derivatives algorithm in a bit of a hurry and this example won't work too well for verifying it's fully distortion free. I guess we might also need a slope for the pixel size or use `w` instead of `z` to become 100% correct. In either case, using derivatives always bears some potential for artifacts at the edges, requires a WebGL extension, and the quality of the derivatives is allowed to vary across WebGL implementations. If I remember correctly, the extension also defines a parameter to switch between "fast", "nice" and "default". Is it set anywhere?

I wonder whether anyone has ever taken the keen step to retire surface normals in favor of quaternions. It's very offbeat, but could yield an improvement in visual fidelity for all kinds of anisotropic shading techniques. It should be quite feasible already and computational power increases much more rapidly than memory bandwidth...

File: examples/js/ShaderSkin.js (hunk 1 above)
Comment (tschw) [inline examples/js/ShaderSkin.js]: This one would have been `viewDir`. I was torn between consistency and clarity and opted for the latter, figuring a reader might not necessarily be familiar with the in-core shading code. I would normally associate the "view direction" with a vector opposite to the "viewer direction". I have no strong opinion on it, just noting that it was done on purpose.

While talking about consistency: Does the inconsistent specular computation for hemispherical light sources in this file actually make sense?

**Comment by mrdoob** [comment]
>  If I remember correctly, the extension also defines a parameter to switch between "fast", "nice" and "default". Is it set anywhere?

Never heard of that...

**Comment by mrdoob** [comment]
I'm still getting an error message here. I wonder if it's just me...?

screen shot 2015-09-06 at 22 43 26

**Comment by arose** [comment]
> If I remember correctly, the extension also defines a parameter to switch between "fast", "nice" and "default". Is it set anywhere?

Only for OpenGL https://www.opengl.org/registry/specs/ARB/derivative_control.txt

**Comment by arose** [comment]
> I'm still getting an error message here. I wonder if it's just me...?

Same here, but only on Chrome 44, FF 42 is fine.

Seems https://github.com/jbouny/ocean/issues/7 Chrome gets sometimes picky about reading/writing from the same texture.

File: examples/js/ShaderSkin.js (hunk 1 above)
Comment (WestLangley) [inline examples/js/ShaderSkin.js]: > Does the inconsistent specular computation for hemispherical light sources in this file actually make sense?

Well, hemisphere lights are a model for ambient light, and typically do not have a specular contribution of that sort. My intention is to remove the specular terms from hemisphere lights in `MeshPhongMaterial` and make sure hemi-lights are property attenuated by AO maps.

**Comment by WestLangley** [comment]
> I came up with the derivatives algorithm in a bit of a hurry

Does that mean you do not have a reference for the technique? Sorry, but adding `pixelSize` to the API just seems bizarre to me. Trying to understand your thought process... : - )

**Comment by tschw** [comment]
@mrdoob wrote:
> > I'm still getting an error message here. I wonder if it's just me...?

@arose wrote:
> Same here, but only on Chrome 44, FF 42 is fine. [ ... ] Chrome gets sometimes picky about reading/writing from the same texture.

OK it's an unrelated issue. I guess we'd have to unbind the textures from the samplers / framebuffers. Thanks! 

@tschw wrote:
> > If I remember correctly, the extension also defines a parameter to switch between "fast", "nice" and "default". Is it set anywhere?

@arose wrote:
> Only for OpenGL https://www.opengl.org/registry/specs/ARB/derivative_control.txt

Nope! See https://www.khronos.org/registry/webgl/extensions/OES_standard_derivatives/ 
There is `FRAGMENT_SHADER_DERIVATIVE_HINT_OES` in GL ES and WebGL. Other than in OpenGL you can only set it globally, however.

@WestLangley wrote:
> Does that mean you do not have a reference for the technique?

No, I built it from scratch just yesterday, so #7109 could be verified.

> Sorry, but adding pixelSize to the API just seems bizarre to me. 

There are two ingredients in different units, thus I have to know their relative scale for stuffing them into the same 3D vector to get a proper direction; if the scale was nonuniform, I would end up with a very distorted direction. The relative scale happens to be the view space size of a pixel.

> Trying to understand your thought process... : - )

Be careful what you wish for =D.

**Comment by arose** [comment]
thanks @tschw, didn't knew that, so it would work like this:

```javascript
var gl = renderer.context;
var extension = renderer.extensions.get( "OES_standard_derivatives" );
gl.hint( extension.FRAGMENT_SHADER_DERIVATIVE_HINT_OES, gl.NICEST );  // FASTEST, NICEST, DONT_CARE
```

see: https://chromium.googlesource.com/chromium/deps/webgl/sdk/tests/+/master/conformance/extensions/oes-standard-derivatives.html

**Comment by WestLangley** [comment]
@tschw 

>Does that mean you do not have a reference for the technique?
>>No, I built it from scratch just yesterday

Why did you not use the approach used in `MeshPhongMaterial` for computing the normal?

**Comment by mrdoob** [comment]
> thanks @tschw, didn't knew that, so it would work like this:

> ```javascript
var gl = renderer.context;
var extension = renderer.extensions.get( "OES_standard_derivatives" );
gl.hint( extension.FRAGMENT_SHADER_DERIVATIVE_HINT_OES, gl.NICEST );  // FASTEST, NICEST, DONT_CARE
```

Just tried this, doesn't seem to make much difference. At least in my hardware.

**Comment by arose** [comment]
I am no expert in this but maybe the approach from http://stackoverflow.com/a/5261402/1435042 to get the tangent space basis helps. I see no apparent artifacts with that.

In `ShaderSkin` use this:

```javascript
"// compute derivations of the view position",
"vec3 p_dx = dFdx( vViewPosition );",
"vec3 p_dy = dFdy( vViewPosition );",
"// compute derivations of the UV coordinate",
"vec2 tc_dx = dFdx( vUv );",
"vec2 tc_dy = dFdy( vUv );",
"// compute initial tangent and bi-tangent",
"vec3 t = normalize( tc_dy.y * p_dx - tc_dx.y * p_dy );",
"vec3 b = normalize( tc_dy.x * p_dx - tc_dx.x * p_dy ); // sign inversion",
"// get new tangent from a given mesh normal",
"vec3 n = normalize( vNormal );",
"vec3 x = cross( n, t );",
"t = cross( x, n );",
"t = normalize( t );",
"// get updated bi-tangent",
"x = cross( b, n );",
"b = cross( n, x );",
"b = normalize( b );",
"mat3 tsb = mat3( t, b, n );",
```

instead of

```javascript
"vec2 uz = vec2( vUv.x, vViewPosition.z );",
"vec2 uzDx = dFdx( uz ), uzDy = dFdy( uz );",
"vec2 tangent2D = normalize( vec2( uzDx.x, uzDy.x ) );",
"vec2 zVec2D = vec2( uzDx.y, uzDy.y );",
"vec3 tangent = vec3( tangent2D * uPixelSize, dot( tangent2D, zVec2D ) );",
"vec3 binormal = normalize( cross( vNormal, tangent ) );",
"tangent = cross( binormal, vNormal );",
"mat3 tsb = mat3( tangent, binormal, vNormal );",
```

**Comment by tschw** [comment]
@WestLangley wrote:
> Why did you not use the approach used in MeshPhongMaterial for computing the normal?

I think that's what I do, but the tangent is not quite so easy to get to: The position won't give it away, the UVs do. To make things worse, we only get a 2D vector with an arbitrary (= unrelated to anything that can help us to figure out Z) length.

It's a little more work than finding the normal. I already explained how my approach works, and it's what I came up with in a couple of minutes. In case you have a simpler solution, let's use that one (there's #7126 that gets rid of the uniform, but it doubles the number of derivatives).

However, in order to work with partial derivatives, and to make any sense of the provided information, it's crucial to know the definition and the geometric properties of the [gradient](https://en.wikipedia.org/wiki/Gradient#Motivation), and that dotting this vector with a unit direction vector gives you the [directional derivative](https://en.wikipedia.org/wiki/Directional_derivative#Definition) (e.g. the slope of a height field) along an arbitrary direction. In GLSL:

```glsl
vec2 gradientOfScalar = vec2( dFdx( scalar ), dFdy( scalar ) );
// tells us in which direction value grows most and how quickly it does

vec2 slopeInDirection = dot( gradientOfScalar, direction );
```

Without knowing these concepts, it's probably as hard to understand as all the other stuff without knowing transforms, vector products, and what else you may need. Once known, it's really no rocket science, just more or less simple geometric calculations as done all over the place.

@mrdoob wrote:
> [... `FRAGMENT_SHADER_DERIVATIVE_HINT_OES` ...]
> Just tried this, doesn't seem to make much difference. At least in my hardware.

It's a hint and the implementation is free to ignore it.

@arose wrote:
> maybe the approach from http://stackoverflow.com/a/5261402/1435042 to get the tangent space basis helps. 

Phew... As noted in the comments, that code is really quite heavy, calculating 10 vs. 4 derivatives. Also, the orthogonalization is incomplete (tangent and binormal are not orthogonalized in respect to each other).

> I see no apparent artifacts with that.

Does it remove any artifacts? Or does it just give similar output?

I'm asking, because I realize when comparing the code, both use different handedness: Maybe I misread the original vertex shader code, or maybe the rest of the code assumes a left-handed space. Also, I could have been more conservative about renormalizing the varyings - the original code didn't bother at all. Also you'd have to use `-vViewPosition` to undo the negation in the vertex shader, I guess.

My variant using a compatible space + conservative normalization (4 derivatives, needs pixelSize):

```glsl
vec2 uz = vec2( vUv.x, vViewPosition.z );
vec2 uzDx = dFdx( uz ), uzDy = dFdy( uz );
vec2 tangent2D = normalize( vec2( uzDx.x, uzDy.x ) );
vec2 zGradient2D = vec2( uzDx.y, uzDy.y );
vec3 tangent = vec3( tangent2D * uPixelSize, dot( zGradient2D, tangent2D ) );
vec3 normal = normalize( vNormal );
vec3 binormal = normalize( cross( tangent, normal ) );
tangent = cross( normal, binormal ); // no normalization required, here
mat3 tsb = mat3( tangent, binormal, normal );
````

Hybrid variant; the tangent is derived based on V, exactly as the post on stackoverflow (now down to 8 derivatives):

```glsl
vec4 posAndV = vec4( vViewPosition, vUv.y );
vec4 posAndV_dx = dFdx( posAndV ),  posAndV_dy = dFdy( posAndV );
vec3 tangent = posAndV_dy.w * posAndV_dx.xyz - posAndV_dx.w * posAndV_dy.xyz;
// [ vec3 normal = normalize( vNormal ); ... see above ]
```

Less obscure hybrid, using U and removing the counter-clockwise 2D rotation, implemented in #7126.

```glsl
vec4 posAndU = vec4( vViewPosition, vUv.x );
vec4 posAndU_dx = dFdx( posAndU ),  posAndU_dy = dFdy( posAndU );
vec3 tangent = posAndU_dx.w * posAndU_dx.xyz + posAndU_dy.w * posAndU_dy.xyz;
// [ vec3 normal = normalize( vNormal ); ... see above ]
```

If we want to use any of it in-core (including code from somewhere on the net) we first need a better test case to really get it right and to compare the quality and the performance of the methods.

@mrdoob wrote:
> Oh?

Interestingly, the post on stackoverflow also suggests considering quaternions instead of surface normals...

**Comment by arose** [comment]
Thanks @tschw for taking the time to explain it some more!

> Does it remove any artifacts? Or does it just give similar output?

With your old code I saw some flat-shaded regions:

old


With the code from #7126 they are gone (as they were with the SO code):

new

**Comment by WestLangley** [comment]
@tschw 

You seem to be under the assumption that I am asking you questions because I do not understand the mathematics. That is not true. I am asking you questions because I do not understand your motivation.

As I implied earlier, I do not think it is a good idea to require the user to specify "the view space size of a pixel" (a.k.a. `pixelSize`) in order for a TSB matrix to be computed correctly.

EDIT: Just noticed #7126. Thank you!

**Comment by arose** [comment]
How does the approach used here relates to the bump mapping in the core library, if I may ask? (I hope this question makes sense, otherwise, sorry for the noise).

Bump mapping example: http://threejs.org/examples/#webgl_materials_bumpmap

Shader code: https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/bumpmap_pars_fragment.glsl

**Comment by tschw** [comment]
@arose

Thanks for the images. 

The artifacts look upside down... Tidying signs / handedness of the space in fact makes them disappear, however, the new code gives more stable output when rotating the object, so doubling the derivatives was well-invested and fragment shader uniforms are a scarce resource.

> How does the approach used here relates to the bump mapping in the core library, if I may ask?

For normal mapping, the normals are encoded within a texture in respect to the surface. Their orientation is defined by the surface normal and the UVs. We therefore derive an orthonormal 3x3 matrix that rotates the normal from the texture to view space (as always, the columns are the unit vectors of the source space expressed in the target space).

For bump mapping, we differentiate over a texture image in respect to screen space. Note that bump maps do not encode an orientation, just height values: The orientation of the data therefore arises from direction of the differentiation. Now we need the orientation of the screen axes projected back onto the surface in order to change the normal. These have some similarities with tangent and binormal, but they are defined on the fly, by the (inverse) screen space projection of the surface - independent from the UVs. Also note that the code you are looking at is really quite nifty; it is derived by solving the linear system that inverts the transform, but the way it's written you can also interpret it geometrically.

@WestLangley 

> You seem to be under the assumption that I am asking you questions because I do not understand the mathematics. That is not true.

I do not make such assumptions, although I figured you might - it's less frequently used stuff compared to, say, vector products after all. Please note the slight difference, and that it was not meant to offend you in any way.

> As I implied earlier, I do not think it is a good idea to require the user to specify "the view space size of a pixel" (a.k.a. pixelSize) in order for a TSB matrix to be computed correctly.

The way I look at it, the user of a custom shader has to pass in whatever information required for efficient rendering. Sweet interfaces are nice to have, but a rather weak argument at this level. For example, this shader requires a texture that serves as a LUT for the Beckmann spectral distribution (a probably outdated optimization, BTW).

Also, your message suggests you knew a better solution all along. So why, instead of making weird comments about the interface, didn't you just provide it?

Whatever... Now the example looks better than ever before :-).

**Comment by WestLangley** [comment]
@tschw

> So why, instead of making weird comments about the interface, didn't you just provide it?

Gee, I do not think my comments were weird.

Rather than imposing a solution on you, I felt it would be more respectful to ask you to further explain your approach. Plus I learn things that way. : - )

**Pull request merged**